### DISCUSSION TITLE
Finding branch upstream via rev-parse or config?

### DISCUSSION THREAD

**Discussion by eamodio** [git-for-windows/git#2981] [General]
VS Code currently uses `rev-parse --symbolic-full-name <branch-name>@{u}` to get the current upstream. This fails in an edge case where the repo is empy but has an upstream branch configured. So we have an open PR to fix this edge case by looking up the branch config. And looking for the remote and merge properties.

IMO, it sounds like a decent fallback, but I would still use `rev-parse` first -- any thoughts on that or is there a better approach?

RE: https://github.com/microsoft/vscode/pull/114163

**Comment by derrickstolee** [comment]
I briefly tried creating a Git test case that mimics your scenario:

```diff
diff --git a/t/t1507-rev-parse-upstream.sh b/t/t1507-rev-parse-upstream.sh
index dfc0d96d8a..75e159ceaa 100755
--- a/t/t1507-rev-parse-upstream.sh
+++ b/t/t1507-rev-parse-upstream.sh
@@ -129,6 +129,16 @@ test_expect_success 'checkout -b new my-side@{u} forks from the same' '
 )
 '
 
+test_expect_success 'rev-parse --symbolic works in empty repo' '
+	git init empty &&
+	git -C empty remote add origin "file://$(pwd)" &&
+	git -C empty config branch.test.remote origin &&
+	git -C empty config branch.test.remote refs/heads/side &&
+	echo side >expect &&
+	git -C empty rev-parse --symbolic test@{u} >actual&&
+	test_cmp expect actual
+'
+
 test_expect_success 'merge my-side@{u} records the correct name' '
 (
 	cd clone &&
```

This test only configures the upstream, but the branch doesn't actually exist. The `rev-parse` command fails with `fatal: no such branch: 'test'`. This matches the scenario in the issue you linked.

Since the repo is empty, I would expect the ref to not exist, so `rev-parse` is reacting sanely here. It probably _shouldn't_ report details for a ref that doesn't actually exist.

I think your fallback is a decent option. It has an issue with "what does 'sync' mean for Git?" It's not an action that corresponds directly to a `git fetch` or `git pull`, so these kinds of corner cases will come up.

**Reply by eamodio** [reply]
Thank you! Yeah, I'm not a fan of "sync" either -- too much magic for me. I prefer a guided but 2-step, pull and then push.
